File: 1927-sum-game/1927-sum-game.cpp
class Solution {
public:
    bool sumGame(string num) {
        int n = num.size();

        int sumLeft = 0;
        int sumRight = 0;

        int qLeft = 0;
        int qRight = 0;

        for (int i = 0; i < n / 2; i++) {
            if (num[i] == '?')
                qLeft++;
            else
                sumLeft += num[i] - '0';
        }

        for (int i = n / 2; i < n; i++) {
            if (num[i] == '?')
                qRight++;
            else
                sumRight += num[i] - '0';
        }

        // If the number of ? on both sides is equal
        if (qLeft == qRight) {
            return sumLeft != sumRight;
        }

        // Difference in number of ? must be even for Bob
        // to potentially balance the sums.
        int qDiff = qLeft - qRight;
        int sumDiff = sumLeft - sumRight;

        return sumDiff * 2 != -9 * qDiff;
    }
};

File: 1927-sum-game/1927-sum-game.js
/**
 * @param {string} num
 * @return {boolean}
 */
var sumGame = function(num) {
    const n = num.length;

    let sumLeft = 0;
    let sumRight = 0;

    let qLeft = 0;
    let qRight = 0;

    for (let i = 0; i < n / 2; i++) {
        if (num[i] === '?') {
            qLeft++;
        } else {
            sumLeft += Number(num[i]);
        }
    }

    for (let i = n / 2; i < n; i++) {
        if (num[i] === '?') {
            qRight++;
        } else {
            sumRight += Number(num[i]);
        }
    }

    if (qLeft === qRight) {
        return sumLeft !== sumRight;
    }

    const qDiff = qLeft - qRight;
    const sumDiff = sumLeft - sumRight;

    return sumDiff * 2 !== -9 * qDiff;
};